echopype/echodata/combine.py

```diff
@@ -435,14 +446,296 @@ def _check_echodata_channels(
+
+    Parameters
+    ----------
+    ds_list: list of xr.Dataset
+        List of Datasets to be combined
+    ed_group: "Vendor_specific"
+        The name of the ``EchoData`` group being combined,
+        this is only works for "Vendor_specific" group.
+    ds_append_dims: set
+        A set of datasets append dimensions
+
+    Returns
+    -------
+    None
+
+    Raises
+    ------
+    ValueError
+        If ``ed_group`` is not ``Vendor_specific``.
+    RuntimeError
+        If non identical filter parameters is found.
+    """
+    if ed_group != "Vendor_specific":
+        raise ValueError("Group must be `Vendor_specific`!")
+
+    if len(ds_append_dims) > 0:
+        # If there's a dataset appending dimension, drop for comparison
+        # of the other values... everything else should be identical
+        ds_list = [ds.drop_dims(ds_append_dims) for ds in ds_list]
+
+    it = iter(ds_list)
+    # Init as identical, must stay True.
+    is_identical = True
+    dataset = next(it)
+    for next_dataset in it:
+        is_identical = dataset.identical(next_dataset)
+        if not is_identical:
+            raise RuntimeError(
+                f"Non identical filter parameters in {ed_group} group. " "Objects cannot be merged!"
+            )
+        dataset = next_dataset
+
+
+def _merge_attributes(attributes: List[Dict[str, str]]) -> Dict[str, str]:
+    """
+    Merges a list of attributes dictionary
+
+    Parameters
+    ----------
+    attributes : list of dict
+        List of attributes dictionary
+        E.g. [{'attr1': 'val1'}, {'attr2': 'val2'}, ...]
+
+    Returns
+    -------
+    dict
+        The merged attribute dictionary
+    """
+    merged_dict = {}
+    for attribute in attributes:
+        for key, value in attribute.items():
+            if value == "" and key not in merged_dict:
+                merged_dict[key] = value
+            elif value != "":
+                merged_dict[key] = value
+    return merged_dict
+
+
+def _capture_prov_attrs(
+    attrs_dict: Dict[str, List[Dict[str, str]]], echodata_filenames: List[str], sonar_model: str
+) -> xr.Dataset:
+    """
+    Capture and create provenance dataset,
+    from the combined attribute values.
+
+    Parameters
+    ----------
+    attrs_dict : dict of list
+        Dictionary of attributes for each of the group.
+        E.g. {'Group': [{'attr1': 'val1'}, {'attr2': 'val2'}, ...]}
+    echodata_filenames : list of str
+        The filenames of the echodata objects
+    sonar_model : str
+        The sonar model
+
+    Returns
+    -------
+    xr.Dataset
+        The provenance dataset for all the attribute values
+
+    """
+    index_keys = [ED_GROUP, ED_FILENAME]
+    df_list = []
+    for group, attributes in attrs_dict.items():
+        df = pd.DataFrame.from_records(attributes)
+        df.loc[:, ED_FILENAME] = echodata_filenames
+        df[ED_GROUP] = group
+        df = df.set_index(index_keys)
```

Comment:
By setting index for both `ED_FILENAME` and `ED_GROUP` the resulting xarray data variables have both of these as dimensions, like below:
<img width="523" alt="Screenshot 2023-05-30 at 10 44 25 AM" src="https://github.com/OSOceanAcoustics/echopype/assets/15334215/f7e9f7b2-ec39-4e23-a4d0-a42c0904ba9c">
<img width="492" alt="Screenshot 2023-05-30 at 10 44 35 AM" src="https://github.com/OSOceanAcoustics/echopype/assets/15334215/3518de5e-69f6-4ec1-b8f2-1544cf50418f">

I found this to be a little confusing? It seems that we can omit the `ED_GROUP` dimension and still append all the attributes?

I assume this is for preserving the "origin" of each attribute. If that is the case, is there another mechanism to store this info, such as in the attributes of the constructed data variables?

Reply:
I didn't even think about storing the group name in each of the data variables! I think this is a really great idea. I'll try that. 😄 Thank you!!!